Add TS GraphQL tutorial (#1425)
encoredev/encore: docs/tutorials/graphql.mdx

</LangTabPanel>
<LangTabPanel>

In this tutorial we will build a GraphQL API using [Apollo](https://www.apollographql.com/docs/apollo-server/) and Encore.ts.

The final code will look like this:

<div className="not-prose my-10">
 <Editor projectName="graphqlTS" />
</div>

<Callout type="info">

To make it easier to follow along, we've laid out a trail of croissants to guide your way.
Whenever you see a 🥐 it means there's something for you to do.

</Callout>

## 1. Create your Encore application

🥐 Create a new application by running `encore app create` and select `Empty app` as the template.

> If this is the first time you're using Encore, you'll be asked if you wish to create a free account. This is needed when you want Encore to manage functionality like secrets and handle cloud deployments (which we'll use later on in the tutorial).

Reviewer: ```suggestion
If this is the first time you're using Encore, you'll be asked if you wish to create a free account. This is optional, but is needed when you want Encore to manage functionality like secrets and handle cloud deployments (which we'll use later on in the tutorial).
```

      }
    }

    // More on how to use executeHTTPGraphQLRequest: https://www.apollographql.com/docs/apollo-server/integrations/building-integrations/
    const httpGraphQLResponse = await server.executeHTTPGraphQLRequest({
      httpGraphQLRequest: {
        headers,
        method: req.method!.toUpperCase(),
        body: await json(req),
        search: new URLSearchParams(req.url ?? "").toString(),
      },
      context: async () => {
        return { req, res };
      },
    });

    for (const [key, value] of httpGraphQLResponse.headers) {
      res.setHeader(key, value);
    }
    res.statusCode = httpGraphQLResponse.status || 200;

    if (httpGraphQLResponse.body.kind === "complete") {
      res.end(httpGraphQLResponse.body.string);
      return;
    }

    for await (const chunk of httpGraphQLResponse.body.asyncIterator) {
      res.write(chunk);
    }
    res.end();
  },
);
```

This creates an [Raw API endpoint](https://encore.dev/docs/ts/primitives/raw-endpoints) available on `/graphql`. In the endpoint we use ApolloServer to handle the GraphQL queries and mutations. We then return the response to the client.

> If we were to use another GraphQL library then Apollo the concept still would be the same:

Reviewer: ```suggestion
If we were to use another GraphQL library other than Apollo, the concept would still be the same:
```

1. Take client requests with a Raw endpoint.
2. Pass along the request and response objects to the GraphQL library of your choice.
3. Use the library to handle the GraphQL queries and mutations.
4. Return the GraphQL response from the Raw endpoint.

## 6. Trying it out

With that, the GraphQL API is done!

🥐Try it out by running `encore run` and opening [https://studio.apollographql.com/sandbox](https://studio.apollographql.com/sandbox) in your browser. Set http://localhost:4000/graphql as your endpoint URL. You should now be able to read the schema and execute queries.

Enter the query:
```graphql
mutation AddBook {
  addBook(author: "J.R.R. Tolkien", title: "The Hobbit") {
    success
    message
    code
  }
}
```

Now try the GetBooks query:

```graphql
query GetBooks {
  books {
    author
    title
  }
}
```

> And you should see the "The Hobbit" in the list of books.

Reviewer: ```suggestion
And you should now see the "The Hobbit" in the list of books.
```

> 🥐 Try opening the Local Development Dashboard at [http://localhost:9400](http://localhost:9400) and view the traces that got generated when calling your GraphQL API.

Reviewer: ```suggestion
🥐 Try opening the Local Development Dashboard at [http://localhost:9400](http://localhost:9400) and view the traces that were generated when calling your GraphQL API.
```

> ## 7. Deploy to the cloud
> 
> 🥐 Push your changes and deploy your application to Encore's free development cloud by running:
> 
> ```shell
> $ git add -A .
> $ git commit -m 'Initial commit'
> $ git push encore
> ```

Reviewer: ```suggestion
## 7. Deploy to the cloud

You can see how to configure and deploy your application in the [self-hosting documentation](/docs/how-to/self-host), or you can use Encore Cloud Platform to automate infrastructure provisioning. More on that below:

🥐 Push your changes and deploy your application to Encore's free development cloud by running:

```shell
$ git add -A .
$ git commit -m 'Initial commit'
$ git push encore
```
```
